### DISCUSSION TITLE
Consider adopting `uwsm` for streamlined session management

### DISCUSSION THREAD

**Discussion by e-tho** [YaLTeR/niri#633] [Ideas]
`niri` could benefit by either prioritizing [`uwsm`](https://github.com/Vladimir-csp/uwsm) for session management or fully delegating this responsibility to it.

`uwsm` effectively addresses many issues related to systemd integration in a clean and universal way.

The specifics of different compositors are handled via [plugins](https://github.com/Vladimir-csp/uwsm/tree/master/uwsm-plugins), in the form of small, easy-to-maintain shell scripts.

**Comment by YaLTeR** [comment]
What benefit would this be to niri? We already integrate systemd, it works fine.

**Reply by e-tho** [reply]
This would [fix](https://github.com/sodiboo/niri-flake/issues/348) the systemd integration on NixOS. Moreover, I like the idea of offloading this responsibility from the compositor to a specific wrapper, as it could lead to a simpler and more efficient setup.

**Reply by YaLTeR** [reply]
> This would fix the systemd integration on NixOS.

As far as it appears to me, there's some NixOS bug that causes this.

> Moreover, I like the idea of offloading this responsibility from the compositor to a specific wrapper, as it could lead to a simpler and more efficient setup.

Niri systemd integration is already about as simple and efficient as it could be though.

**Reply by e-tho** [reply]
> As far as it appears to me, there's some NixOS bug that causes this.

Indeed, this is an issue related to NixOS that has existed for years, and the solution that seems to be getting the most attention and support is `uwsm`. I just saw that its module has been [merged](https://github.com/NixOS/nixpkgs/pull/335625) two days ago. I'll test it when it [lands](https://nixpk.gs/pr-tracker.html?pr=335625) on the unstable branch.

> Niri systemd integration is already about as simple and efficient as it could be though.

I think that official support for `uwsm` in Niri would be beneficial for all distributions, and I see it as a clean way to separate responsibilities. Plus, by handling the session management in a separate project, Niri can benefit from the ongoing contributions made to the wrapper.

**Reply by e-tho** [reply]
The issue persists in when niri is run through UWSM on NixOS, here's the result:
```console
╰─ sudo journalctl -xb | rg uwsm
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]: Preparing environment for niri...
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]: Exporting variables to systemd user manager:
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   DOCKER_HOST
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   EDITOR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   GTK_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   INFOPATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LESSKEYIN_SYSTEM
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LESSOPEN
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LIBEXEC_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIXPKGS_CONFIG
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_PROFILES
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_USER_PROFILE_DIR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   PAGER
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   QTWEBKIT_PLUGIN_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   SSH_ASKPASS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   TERM
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   TERMINFO_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XCURSOR_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_BACKEND
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_CONFIG_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_CURRENT_DESKTOP
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_DATA_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_MENU_PREFIX
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_DESKTOP
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_ID
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_TYPE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_VTNR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   __ETC_PROFILE_DONE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   __NIXOS_SET_ENVIRONMENT_DONE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]: Variables marked for cleanup from systemd user manager on stop:
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   DISPLAY
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   DOCKER_HOST
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   EDITOR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   GTK_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   INFOPATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LANG
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LESSKEYIN_SYSTEM
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LESSOPEN
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   LIBEXEC_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIXPKGS_CONFIG
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_PROFILES
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   NIX_USER_PROFILE_DIR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   PAGER
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   QTWEBKIT_PLUGIN_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   SSH_ASKPASS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   TERM
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   TERMINFO_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   WAYLAND_DISPLAY
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XCURSOR_PATH
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XCURSOR_SIZE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XCURSOR_THEME
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_BACKEND
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_CONFIG_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_CURRENT_DESKTOP
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_DATA_DIRS
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_MENU_PREFIX
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_DESKTOP
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_ID
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_SESSION_TYPE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   XDG_VTNR
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   __ETC_PROFILE_DONE
Aug 29 19:13:44 nixos uwsm_env-preloader[2163]:   __NIXOS_SET_ENVIRONMENT_DONE
Aug 29 19:13:44 nixos uwsm_niri[2211]: Starting: /run/current-system/sw/bin/niri...
Aug 29 19:13:44 nixos uwsm_niri[2211]: 2024-08-29T17:13:44.461718Z  INFO niri: starting version unstable 2024-08-26 (commit 598b27f83c4d4dd72c5cfbd06ca60b12e2833429)
Aug 29 19:13:44 nixos uwsm_niri[2211]: 2024-08-29T17:13:44.469504Z DEBUG niri_config: loaded config from "/home/etho/.config/niri/config.kdl"
Aug 29 19:13:44 nixos uwsm_niri[2211]: 2024-08-29T17:13:44.944623Z  INFO niri::backend::tty: using as the render node: "/dev/dri/renderD128"
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.069355Z DEBUG niri::backend::tty: device added: 57857 "/dev/dri/card1"
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.380679Z DEBUG niri::backend::tty: this is the primary node
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.380692Z DEBUG niri::backend::tty: this is the primary render node
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.428766Z DEBUG niri::backend::tty: device changed: 57857
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.433015Z DEBUG niri::backend::tty: connecting connector: DP-1
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.433042Z DEBUG niri::backend::tty: picking mode: Mode { name: "2560x1440", clock: 645000, size: (2560, 1440), hsync: (2568, 2600, 2640), vsync: (1446, 1454, 1480), hskew: 0, vscan: 0, vrefresh: 165, mode_type: ModeTypeFlags(DRIVER) }
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.433082Z DEBUG niri::backend::tty: set max bpc to 8
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.440093Z DEBUG niri::niri: putting output DP-1 at x=0 y=0
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.440148Z  INFO niri: listening on Wayland socket: wayland-1
Aug 29 19:13:45 nixos uwsm_niri[2211]: 2024-08-29T17:13:45.440154Z  INFO niri: IPC listening on: /run/user/1000/niri.wayland-1.2211.sock
Aug 29 19:14:50 nixos sudo[2897]:     etho : TTY=pts/0 ; PWD=/home/etho ; USER=root ; COMMAND=/run/current-system/sw/bin/journalctl -xb -u uwsm
╰─ systemctl --user status swayidle.service
○ swayidle.service - Idle manager for Wayland
     Loaded: loaded (/home/etho/.config/systemd/user/swayidle.service; enabled; preset: enabled)
     Active: inactive (dead)
  Condition: start condition unmet at Thu 2024-08-29 19:13:44 CEST; 4min 5s ago
             └─ ConditionEnvironment=WAYLAND_DISPLAY was not met
       Docs: man:swayidle(1)

Aug 29 19:13:44 nixos systemd[2039]: Idle manager for Wayland was skipped because of an unmet condition check (ConditionEnvironment=WAYLAND_DISPLAY).
```
I also tried passing `niri-session` to UWSM but the services don't start either and the session crashes after a few seconds.

**Reply by Milner39** [reply]
I also have this issue trying to use Niri on NixOS

**Reply by e-tho** [reply]
@Milner39, Here's the solution I've been using for over a year. Works flawlessly.

```nix
{
  pkgs,
  inputs,
  lib,
  ...
}:
let
  niriInstance = lib.getExe (
    pkgs.writeShellScriptBin "niri-instance" ''
      /run/current-system/sw/bin/niri --session
    ''
  );
in
{
  environment.systemPackages = with inputs; [
    niri.packages.${pkgs.system}.default
    xwayland-satellite.packages.${pkgs.system}.default
  ];

  programs = {
    uwsm = {
      enable = true;
      package = pkgs.uwsm;
      waylandCompositors = {
        niri = {
          prettyName = "Niri";
          comment = "Niri compositor managed by UWSM";
          binPath = niriInstance;
        };
      };
    };
  };
}
```

The systemd integration (unit ordering) issue on NixOS was due to Home Manager services being incorrectly written. See:
* https://github.com/Vladimir-csp/uwsm/issues/40
* https://github.com/sodiboo/niri-flake/issues/348
* https://github.com/nix-community/home-manager/pull/6253

**Reply by Milner39** [reply]
Pwahhh you legend, thanks!

**Reply by quantumfate** [reply]
> Niri systemd integration is already about as simple and efficient as it could be though.

this is going against the unix philosophy

**Reply by cardboardcpu** [reply]
> > Niri systemd integration is already about as simple and efficient as it could be though.
> 
> this is going against the unix philosophy 

The "Unix philosophy", while having some kernel of useful information among its precepts, it's basically an outdated way to view modern software. Even harmful when taken as gospel.

Propaganda aside, I would argue that, for most usecases, UWSM is only useful in situations where the compositor lacks, partially or integrally, good SystemD integration. So as long as Niri provides good support it's just redundant.

Also, UWSM is _slow_.

I run Niri with UWSM btw. :D

**Reply by zefr0x** [reply]
Two cases I can think about where UWSM might be beneficial in niri.

First, when a user already have a setup for UWSM with environment variables used in other compositor, they don't need to redefine them again.

Second, if a user is relying on UWSM (with the `uwsm app` command) for launching applications as services in systemd scopes (e.g. `app-graphical.slice`, `background-graphical.slice`, `session-graphical.slice`) prioritising resources for some desktop applications.

Note that I'm against removing current systemd integration in favour for UWSM. Also, I don't think that niri need to support UWSM, but rather just allow it to be used without errors as another option.

**Reply by quantumfate** [reply]
> The "Unix philosophy", while having some kernel of useful information among its precepts, it's basically an outdated way to view modern software. Even harmful when taken as gospel.

While I partially agree it's also worth considering the additional cost of maintaining an aspect of your software. This might be trivial in regards to Niri and its Systemd coverage but looking at Hyprland, its growing ecosystem almost creates vendor lock in. [Hyprshutdown](https://github.com/hyprwm/hyprshutdown) for instance is completely useless when you wrap Hyprland in UWSM and it also provides the benefit of launching your applications outside of the compositor scope.

This is just personal preference but I like to be completely independent and delegate scopes and tasks to software that was meant to solve them. As for the overall health of Niri, Hyprland, etc. it's perfectly fine and I absolutely agree with the notion solving these problems themselves to tackle the the influx of new inexperienced Linux users and providing a smooth first experience.

For more experienced Linux users however, I'd like to see the ability to opt out of certain features.
